### PR TITLE
Deduplication Batch Index Hook

### LINKED ISSUE #32: Inverse Index Lookup Hook

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
Close #32 

High-level usage:
```python

# Just do inference on unique ids
z = model(batch.unique_nids) 

# Use *_idx fields to lookup the relevant embeddings
src_z = z[batch.src_idx]
dst_z = z[batch.dst_idx]

# These are optional fields that are added if using negative sampler or nbr sampler, respectively
neg_z = z[batch.neg_idx] if hasattr(batch, 'neg') else None
nbr_z = [z[idx] for idx in batch.nbr_nids_idx] if hasattr(batch, 'nbr_nids') else None

# If you got custom hooks and want to just get the embeddings for some arbitrary node tensors (e.g. tgb)
query_ids: torch.Tensor = ...
z_query_ids = z[batch.nid_to_idx[query_ids]]
```

### CODE REVIEW

File: tgm/hooks.py [hunk 1]
```diff
@@ -95,6 +98,50 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
 
+class DeduplicationHook:
+    requires: Set[str] = set()
+    produces = {
+        'unique_nids',
+        'nid_to_idx',
+        'src_idx',
+        'dst_idx',
+        'neg_idx',
+        'nbr_nids_idx',
```
Comment (shenyangHuang) [inline tgm/hooks.py:109]: why does it produce 'nbr_nids_idx'? We should do Dedup before sampling neighbors right?

File: tgm/hooks.py [hunk 2]
```diff
@@ -95,6 +98,56 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
+        if hasattr(batch, 'neg'):
+            nids.append(batch.neg)
+        if hasattr(batch, 'nbr_nids'):
+            for hop in range(len(batch.nbr_nids)):
+                hop_nids, hop_mask = batch.nbr_nids[hop], batch.nbr_mask[hop].bool()  # type: ignore
+                nids.append(hop_nids[hop_mask])
+
+        all_nids = torch.cat(nids, dim=0)
+        unique_nids = torch.unique(all_nids)
+        max_nid = int(unique_nids.max().item())
```
Comment (Jacob-Chmura) [inline tgm/hooks.py:130]: The dense tensor representation is going to be slightly better than the native python dict, but we should beware that this becomes very memory hungry when the `max_nid` is very large... And we allocate these tensors on each batch...

File: tgm/hooks.py [hunk 3]
```diff
@@ -95,6 +98,56 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
 
+class DeduplicationHook:
+    requires: Set[str] = set()
+    produces = {
+        'unique_nids',
+        'nid_to_idx',
+        'src_idx',
+        'dst_idx',
+        'neg_idx',
+        'nbr_nids_idx',
```
Comment (Jacob-Chmura) [inline tgm/hooks.py:109]: Arguably, from a modularity perspective we could skip these all together and just give the user the `nid_to_idx` and let them lookup whichever indices they want (e.g. how we discussed it would work for tgb negatives). Since the user has `batch.src, batch.dst, batch.neg` they can do `z[batch.nid_to_idx[batch.src]]` themselves.

It gets slightly more complex with neighbours (since you need to mask out invalid neighborhoods, and need to be on the _right hop_. I think this is worth discussing though.

File: examples/nodeproppred/gcn.py
Comment (shenyangHuang) [inline examples/nodeproppred/gcn.py:1]: Looks great

File: tgm/hooks.py (hunk 2 above)
Comment (shenyangHuang) [reply tgm/hooks.py:130]: that is true, however, it is still a 1d tensor so should be reasonably good, definitely can optimize later on.
